stake-pool/program/src/state.rs

```diff
@@ -0,0 +1,102 @@
+//! State transition types
+
+use crate::error::Error;
+use crate::instruction::{unpack, Fee};
+use solana_sdk::{entrypoint::ProgramResult, program_error::ProgramError, pubkey::Pubkey};
+use std::mem::size_of;
+
+/// Initialized program details.
+#[repr(C)]
+#[derive(Clone, Copy, Debug, Default, PartialEq)]
+pub struct StakePool {
+    /// Owner authority
+    /// allows for updating the staking authority
+    pub owner: Pubkey,
+    /// Deposit authority nonce
+    /// for `create_program_address(&[state::StakePool account, "deposit"])`
+    pub deposit_nonce: u8,
+    /// Withdrawal authority nonce
+    /// for `create_program_address(&[state::StakePool account, "withdrawal"])`
+    pub withdraw_nonce: u8,
+    /// Pool Mint
+    pub pool_mint: Pubkey,
+    /// Owner fee account
+    pub owner_fee_account: Pubkey,
+    /// Pool token program id
+    pub token_program_id: Pubkey,
+    /// total stake under management
+    pub stake_total: u64,
+    /// total pool
+    pub pool_total: u64,
+    /// Fee applied to deposits
+    pub fee: Fee,
+}
+impl StakePool {
+    /// calculate the pool tokens that should be minted
+    pub fn calc_pool_amount(&self, stake_lamports: u64) -> Option<u128> {
+        (stake_lamports as u128)
+            .checked_mul((self.pool_total as u128).checked_add(1)?)?
+            .checked_div((self.stake_total as u128).checked_add(1)?)
+    }
+    /// calculate the fee in pool tokens that goes to the owner
+    pub fn calc_fee_amount(&self, pool_amount: u128) -> Option<u128> {
+        pool_amount
+            .checked_mul((self.fee.numerator as u128).checked_add(1)?)?
```

Comment:
Actually, just noticing this more, this will cause a problem if someone wants a 0 fee